File: packages/sandbox/daemon/routes/tasks.ts
import type { TaskManager, TaskStatus } from "../process/task-manager";
import { sseFormat } from "../events/sse-format";
import { jsonResponse } from "./body-parser";

export async function awaitTaskResponse(
  taskManager: TaskManager,
  id: string,
  opts: { extra?: Record<string, unknown>; timedOutExitCode?: number } = {},
): Promise<Response> {
  const wait = taskManager.finished(id);
  if (!wait)
    return jsonResponse({ error: "task vanished before completion" }, 500);
  const result = await wait;
  const out = taskManager.output(id);
  const exitCode =
    opts.timedOutExitCode !== undefined && result.timedOut
      ? opts.timedOutExitCode
      : result.exitCode;
  return jsonResponse({
    ...opts.extra,
    stdout: out?.stdout ?? "",
    stderr: out?.stderr ?? "",
    exitCode,
    timedOut: result.timedOut,
    truncated: out?.truncated ?? false,
  });
}

export interface TasksDeps {
  taskManager: TaskManager;
}

const VALID_STATUS: ReadonlySet<TaskStatus> = new Set([
  "running",
  "exited",
  "failed",
  "killed",
  "timeout",
]);

/** GET /_decopilot_vm/tasks?status=running,exited */
export function makeTasksListHandler(deps: TasksDeps) {
  return async (req: Request): Promise<Response> => {
    const url = new URL(req.url);
    const statusParam = url.searchParams.get("status");
    const status = statusParam
      ? statusParam
          .split(",")
          .filter((s): s is TaskStatus => VALID_STATUS.has(s as TaskStatus))
      : undefined;
    const tasks = deps.taskManager.list(
      status && status.length > 0 ? { status } : undefined,
    );
    return jsonResponse({ tasks });
  };
}

/** GET /_decopilot_vm/tasks/:id */
export function makeTasksGetHandler(deps: TasksDeps) {
  return async (req: Request): Promise<Response> => {
    const id = idFrom(req, "/tasks/");
    if (!id) return jsonResponse({ error: "missing task id" }, 400);
    const summary = deps.taskManager.get(id);
    if (!summary) return jsonResponse({ error: "task not found" }, 404);
    const out = deps.taskManager.output(id);
    return jsonResponse({
      ...summary,
      stdout: out?.stdout ?? "",
      stderr: out?.stderr ?? "",
      truncated: out?.truncated ?? false,
    });
  };
}

/** POST /_decopilot_vm/tasks/:id/kill[?signal=SIGTERM|SIGKILL] */
export function makeTasksKillHandler(deps: TasksDeps) {
  return async (req: Request): Promise<Response> => {
    const id = idFrom(req, "/tasks/", "/kill");
    if (!id) return jsonResponse({ error: "missing task id" }, 400);
    const url = new URL(req.url);
    const sig = (url.searchParams.get("signal") ?? "SIGTERM") as NodeJS.Signals;
    const ok = deps.taskManager.kill(id, sig);
    if (!ok) return jsonResponse({ error: "task not running" }, 400);
    return jsonResponse({ ok: true });
  };
}

/** POST /_decopilot_vm/tasks/kill-all */
export function makeTasksKillAllHandler(deps: TasksDeps) {
  return async (): Promise<Response> => {
    const count = deps.taskManager.killAll();
    return jsonResponse({ ok: true, killed: count });
  };
}

/** DELETE /_decopilot_vm/tasks/:id */
export function makeTasksDeleteHandler(deps: TasksDeps) {
  return async (req: Request): Promise<Response> => {
    const id = idFrom(req, "/tasks/");
    if (!id) return jsonResponse({ error: "missing task id" }, 400);
    const ok = deps.taskManager.delete(id);
    if (!ok)
      return jsonResponse({ error: "task not found or still running" }, 400);
    return jsonResponse({ ok: true });
  };
}

/** GET /_decopilot_vm/tasks/:id/stream — SSE: replay buffered output, then live. */
export function makeTasksStreamHandler(deps: TasksDeps) {
  return async (req: Request): Promise<Response> => {
    const id = idFrom(req, "/tasks/", "/stream");
    if (!id) return jsonResponse({ error: "missing task id" }, 400);
    const summary = deps.taskManager.get(id);
    if (!summary) return jsonResponse({ error: "task not found" }, 404);

    const stream = new ReadableStream<Uint8Array>({
      start(controller) {
        const send = (event: string, payload: unknown) => {
          try {
            controller.enqueue(sseFormat(event, JSON.stringify(payload)));
          } catch {
            /* controller closed */
          }
        };

        const replay = deps.taskManager.output(id);
        if (replay) {
          if (replay.stdout) send("stdout", { data: replay.stdout });
          if (replay.stderr) send("stderr", { data: replay.stderr });
        }
        if (summary.status !== "running") {
          send("end", {
            status: summary.status,
            exitCode: summary.exitCode,
            timedOut: summary.timedOut,
          });
          controller.close();
          return;
        }

        const unsubscribe = deps.taskManager.subscribe(id, (chunk) => {
          send(chunk.stream, { data: chunk.data });
        });

        void deps.taskManager.finished(id)?.then((result) => {
          send("end", {
            status: result.status,
            exitCode: result.exitCode,
            timedOut: result.timedOut,
          });
          unsubscribe?.();
          try {
            controller.close();
          } catch {
            /* already closed */
          }
        });

        req.signal.addEventListener("abort", () => {
          unsubscribe?.();
          try {
            controller.close();
          } catch {
            /* already closed */
          }
        });
      },
    });

    return new Response(stream, {
      status: 200,
      headers: {
        "Content-Type": "text/event-stream",
        "Cache-Control": "no-cache",
        Connection: "keep-alive",
        "Access-Control-Allow-Origin": "*",
        "X-Accel-Buffering": "no",
      },
    });
  };
}

function idFrom(req: Request, prefix: string, suffix?: string): string | null {
  const url = new URL(req.url);
  const tail = url.pathname.split(prefix)[1];
  if (!tail) return null;
  let id = tail;
  if (suffix && id.endsWith(suffix)) {
    id = id.slice(0, -suffix.length);
  }
  if (!id || id.includes("/")) return null;
  return id;
}

File: packages/sandbox/daemon-go/internal/routes/tasks.go
package routes

import (
	"net/http"
	"strings"
	"syscall"

	"github.com/decocms/studio/sandbox-daemon/internal/events"
	"github.com/decocms/studio/sandbox-daemon/internal/httpx"
	"github.com/decocms/studio/sandbox-daemon/internal/proc"
)

type TasksDeps struct {
	TaskManager *proc.TaskManager
}

func taskIdFrom(path, prefix, suffix string) string {
	idx := strings.Index(path, prefix)
	if idx < 0 {
		return ""
	}
	id := path[idx+len(prefix):]
	if suffix != "" && strings.HasSuffix(id, suffix) {
		id = id[:len(id)-len(suffix)]
	}
	if id == "" || strings.Contains(id, "/") {
		return ""
	}
	return id
}

func TasksList(deps TasksDeps) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		var statuses []string
		if raw := r.URL.Query().Get("status"); raw != "" {
			for _, s := range strings.Split(raw, ",") {
				if proc.ValidStatuses[s] {
					statuses = append(statuses, s)
				}
			}
		}
		httpx.JSON(w, 200, map[string]any{"tasks": deps.TaskManager.List(statuses)})
	}
}

func TasksGet(deps TasksDeps) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		id := taskIdFrom(r.URL.Path, "/tasks/", "")
		if id == "" {
			httpx.Error(w, 400, "missing task id")
			return
		}
		summary, ok := deps.TaskManager.Get(id)
		if !ok {
			httpx.Error(w, 404, "task not found")
			return
		}
		out, _ := deps.TaskManager.Output(id)
		httpx.JSON(w, 200, map[string]any{
			"id": summary.ID, "command": summary.Command, "status": summary.Status,
			"exitCode": summary.ExitCode, "startedAt": summary.StartedAt,
			"finishedAt": summary.FinishedAt, "timedOut": summary.TimedOut,
			"logName": omitEmpty(summary.LogName), "intentional": summary.Intentional,
			"stdout": out.Stdout, "stderr": out.Stderr, "truncated": out.Truncated,
		})
	}
}

func omitEmpty(s string) any {
	if s == "" {
		return nil
	}
	return s
}

func TasksKill(deps TasksDeps) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		id := taskIdFrom(r.URL.Path, "/tasks/", "/kill")
		if id == "" {
			httpx.Error(w, 400, "missing task id")
			return
		}
		sig := syscall.SIGTERM
		if r.URL.Query().Get("signal") == "SIGKILL" {
			sig = syscall.SIGKILL
		}
		if !deps.TaskManager.Kill(id, sig) {
			httpx.Error(w, 400, "task not running")
			return
		}
		httpx.JSON(w, 200, map[string]any{"ok": true})
	}
}

func TasksKillAll(deps TasksDeps) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		count := deps.TaskManager.KillAll()
		httpx.JSON(w, 200, map[string]any{"ok": true, "killed": count})
	}
}

func TasksDelete(deps TasksDeps) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		id := taskIdFrom(r.URL.Path, "/tasks/", "")
		if id == "" {
			httpx.Error(w, 400, "missing task id")
			return
		}
		if !deps.TaskManager.Delete(id) {
			httpx.Error(w, 400, "task not found or still running")
			return
		}
		httpx.JSON(w, 200, map[string]any{"ok": true})
	}
}

func TasksStream(deps TasksDeps) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		id := taskIdFrom(r.URL.Path, "/tasks/", "/stream")
		if id == "" {
			httpx.Error(w, 400, "missing task id")
			return
		}
		summary, ok := deps.TaskManager.Get(id)
		if !ok {
			httpx.Error(w, 404, "task not found")
			return
		}

		h := w.Header()
		h.Set("Content-Type", "text/event-stream")
		h.Set("Cache-Control", "no-cache")
		h.Set("Connection", "keep-alive")
		h.Set("Access-Control-Allow-Origin", "*")
		h.Set("X-Accel-Buffering", "no")
		w.WriteHeader(200)
		flusher, _ := w.(http.Flusher)

		send := func(event string, payload any) {
			w.Write(events.SseFrame(event, payload))
			if flusher != nil {
				flusher.Flush()
			}
		}

		out, _ := deps.TaskManager.Output(id)
		if out.Stdout != "" {
			send("stdout", map[string]string{"data": out.Stdout})
		}
		if out.Stderr != "" {
			send("stderr", map[string]string{"data": out.Stderr})
		}
		if summary.Status != proc.StatusRunning {
			send("end", map[string]any{
				"status": summary.Status, "exitCode": summary.ExitCode, "timedOut": summary.TimedOut,
			})
			return
		}

		chunks := make(chan proc.OutputChunk, 256)
		unsubscribe, ok := deps.TaskManager.Subscribe(id, func(c proc.OutputChunk) {
			select {
			case chunks <- c:
			default:
			}
		})
		if !ok {
			send("end", map[string]any{"status": summary.Status})
			return
		}
		defer unsubscribe()

		finished := make(chan proc.TaskResult, 1)
		go func() {
			result, ok := deps.TaskManager.Finished(id)
			if ok {
				finished <- result
			}
		}()

		for {
			select {
			case c := <-chunks:
				send(c.Stream, map[string]string{"data": c.Data})
			case result := <-finished:
				for {
					select {
					case c := <-chunks:
						send(c.Stream, map[string]string{"data": c.Data})
						continue
					default:
					}
					break
				}
				send("end", map[string]any{
					"status": result.Status, "exitCode": result.ExitCode, "timedOut": result.TimedOut,
				})
				return
			case <-r.Context().Done():
				return
			}
		}
	}
}
